workflow/controller/controller.go

```diff
@@ -236,6 +242,69 @@ func (wfc *WorkflowController) podGarbageCollector(stopCh <-chan struct{}) {
 	}
 }
 
+func (wfc *WorkflowController) workflowGarbageCollector(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case uid := <-wfc.gcWorkflows:
+			logCtx := log.WithField("uid", uid)
+			logCtx.Debug("Performing workflow GC")
+			err := wfc.offloadNodeStatusRepo.Delete(string(uid))
+			if err != nil {
+				logCtx.WithField("err", err).Error("GC failed")
+			}
+		}
+	}
+}
+
+func (wfc *WorkflowController) periodicWorkflowGarbageCollector(stopCh <-chan struct{}) {
+	value, ok := os.LookupEnv("WORKFLOW_GC_PERIOD")
+	periodicity := 5 * time.Minute
+	if ok {
+		var err error
+		periodicity, err = time.ParseDuration(value)
+		if err != nil {
+			log.WithFields(log.Fields{"err": err, "value": value}).Fatal("Failed to parse WORKFLOW_GC_PERIOD")
+		}
+	}
+	log.Infof("Performing periodic GC every %v", periodicity)
+	ticker := time.NewTicker(periodicity)
+	for {
+		select {
+		case <-stopCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			if wfc.offloadNodeStatusRepo.IsEnabled() {
+				log.Info("Performing periodic workflow GC")
+				uids, err := wfc.offloadNodeStatusRepo.ListOldUIDs(wfc.GetManagedNamespace())
+				if err != nil {
+					log.WithField("err", err).Error("Failed to list offloaded nodes")
+				}
+				list, err := wfc.wfclientset.ArgoprojV1alpha1().Workflows(wfc.GetManagedNamespace()).List(metav1.ListOptions{})
```

Comment:
We should use an informer List instead of an API call. Informer lists will not call the API server and it does not error.